### DISCUSSION TITLE
"Select All" functionality needed in column select

### DISCUSSION THREAD

**Discussion by lmaloney** [apache/superset#32779] [Ideas]
### Discussed in https://github.com/apache/superset/discussions/18247

<div type='discussions-op-text'>

<sup>Originally posted by **cccs-rc** February  1, 2022</sup>
Hi all, 

Having recently deployed 1.4, we've discovered that the "Select All" option (pictured below as seen in 1.3) was removed from column SelectControl in the Explore view. It appears to have been a conscious decision made during the work done for [this SelectControl refactoring PR](https://github.com/apache/superset/pull/16510/files). However, our users rely heavily on that "Select All" functionality, so we're hoping this can be brought back in some form or fashion. 

<img src="https://user-images.githubusercontent.com/62034438/152042546-27a0095a-7e9e-448b-9f12-2f8e98698ac7.png" width="300">

Without a "Select All" functionality, users are forced to manually select all the fields they wish to include in a chart. This doesn't scale well when dealing with datasets that have 100+ columns. Our users often want to view such data sets in table charts. While they may ultimately decide to pare the number of fields down to those that are most relevant to their use cases, when users are initially developing charts, it can be critical for them to be able to easily view all the fields/data available in order to identify which fields are most relevant. 

Our users rely on this functionality so heavily that we've already jammed it back into our fork, so we've seen that things can get a little weird in the UI with the new functionality _and_ a Select All option (e.g., when selected fields are moved to the top of the column list, where should the "Select All" button go?). We're hoping we can come up with a design that will allow us to keep the "Select All" functionality while maintaining good UX. 

We're happy to do the work and put in a PR if we can get some insight into what criteria the design should meet. 

Any thoughts/insight would be much appreciated! 

Thanks! 
</div>

**Comment by villebro** [comment]
Thanks for opening the issue @lmaloney. This has been raised by many people since the option was removed. The main motivation for removal is that selecting all columns is a fairly uncommon use case when creating charts - the majority of charts would not easily handle grouping by 100+ columns, both from a visualization perspective, but also from a query performance perspective. However, I understand the use case e.g. in the table chart when using "raw records" mode. Out of curiosity, are you looking to use it exclusively in "raw records" mode in the table chart, or in other charts, too?

**Comment by lmaloney** [comment]
Ah, I _think_ there is a disconnect here....

My use case, as well as "CCC-RC" (I suspect) use case is  indeed the **_data table chart._**

I don't really use the graphing functionality all that much of the charts... I do.. but, 90% of the charts I produce are data tables.  Either raw data, OR summaries of data. 

Unfortunately most of the data tables just have LOTS of columns.

The data is pre-processed, and cached. (the underlying data table is actually a _view_, with **pre-match** results stored in a lookup table that just uses the IDs).

So querying the results (even in the tens of thousands ) takes sub-second time.  Of course the presentation within SuperSet is limited to a few thousand at a time.  Ergo, performance has not been an issue at all.

So, yes, it's primarily "raw" and it's the data table.

To help persuade justifying re-instatement of this feature, I would point out the obvious... that if many people have raised this issue... then it is indeed used quite a bit! :)

Finally, I want to state that I've been using Superset for my application for the past six months, and I love it so much.  I have not experienced any bugs or problems, and I'm looking forward to getting some spare time to contribute to the project.

**Comment by rusackas** [comment]
@michael-s-molina / @cccs-RyanK do you happen to know the state of affairs here, and if/when this particular issue might be resolvable?

**Comment by michael-s-molina** [comment]
@rusackas https://github.com/apache/superset/pull/22084 introduced Select All for synchronous selects with `mode = 'multiple'`. Using Explore with dnd, the Columns popover uses the Select in single mode, which will not enable Select All. If I'm not mistaken, @kasiazjc was working on some prototypes for this issue. I'm not sure if it uses the Select control to resolve it.

**Comment by SharonCastel** [comment]
How can I enable the option for select all? is there a progress that I missed?

**Comment by lmaloney** [comment]
To follow up for any others coming here, you too @SharonCastel , there is a way to do this.... 

The current way to do this, is to create a view/query in the SQL labs.  If you make a custom query in sql labs, you can drag and drop all your columns int to the query. when you save off the query, you can automatically going the chart builder, and all your fields will be in the list for the table/chart to use without having to drag and drop every single one.

**Comment by rusackas** [comment]
Hi all,

I'm not quite sure what to do with this one. Arguably, it's stale and could be closed by that argument. But I know it's still present and important.

Then again it's more of a feature request than a bug... so it should be a Discussion, not an issue. We could close this in favor of https://github.com/apache/superset/discussions/18247 by that argument. 

But... I'd like to see this happen (like many others), as I've wished for it. Does anyone here have the intent to contribute such a feature? @kasiazjc did you wind up with any designs around this perchance?

**Comment by cccs-rc** [comment]
I believe the open questions about design were what stopped our previous contribution. Unfortunately, I don't think we have cycles to try to revive it, but for whoever takes this on, agreement on UI design will be an important piece! :)

**Comment by rusackas** [comment]
This has gone stale for upward of a year now. I'm tempted to close this as stale, but since it's still a legit idea (I wish for it often), I'll leave it open and migrate it to a GitHub Discussion. Contributions/proposals are still welcomed with open arms :D

**Comment by DZHEX** [comment]
This looks like a rare case where open-source software has made a significant regression in terms of usability, usually reserved for the ensh*ttification process of corporate, closed-source software. I work at an environment where we do not (are allegedly not supposed to) have access to SQLLab. Literally any implementation of a "select all" option, even as a non-searchable, nondescript button (column) buried in a list of columns is preferable to current state.
Right now I see a list of metrics and columns on the left, which I can drag into the "Columns" or "Dimensions" field. Maybe there can be a special "select all" entry at the top of the list, which one could drag into the same box, which adds all columns? Something similar for the box that opens when you click in the "Columns" field? Maybe there can be entries like "Select all", "Select none" (in case you selected all/a lot of columns by mistake), "Invert selection" (so you can select ones you don't want and then flip it) that you drag into the box just like entries from the other lists, in each of the "Columns" and "Metrics" lists, which execute their actions when dropped?

I apologize for my somewhat angry attitude in this comment, but I find it very bizarre that this relatively simple and actually useful tool was removed by a conscious decision. As I said - literally any implementation of the option to select all via UI, however difficult for/antagonistic towards the user, is preferable to lack thereof, if it actually allows you to make a table of 100+ columns, without having to select them one-by-one.

I hope that it is brought back soon (the folks running our infra will also have to be nagged for additional months to get the update once it comes out :joy: )
